keymanapp/keyman
# fix(android/engine): Update countSurrogatePairs()

Fixes #6301
This replaces `countSurrogatePairs()` with the code block added in #6274, as it's really only used in `KMManager.performLeftDeletions()`. Notably, `countSurrogatePairs()` no longer includes split surrogate pairs in its count.
This affects the unit tests accordingly.

While cleaning up CharSequenceUtils, also removed unused `adjustCursorPosition()`.

## User Testing
Setup - 
1. On an Android emulator of SDK 21, load the PR build of Keyman for Android
2. In the Keyman app, install the following Keyman keyboards:
    * xinaliq - This uses rota keys 
    * basic_kbdoldit - This uses SMP characters of "Old Italic". Note, the keyboard doesn't include a font, so the Android emulator will display tofu (square boxes). Alternatively, you can install this [test version of basic_kbdoldit](https://darcywong00.github.io/examples/basic_kbdoldit/basic_kbdoldit.kmp) which includes a font for the keyboard.
3. From the Keyman settings, set Keyman as a system keyboard and enable as default.
4. Some of the test steps will ask to verify handling of split-surrogate pairs. You can get some at this test page
https://darcywong00.github.io/examples/custom.html
The smiley emoji is `U+1F600` which consists of surrogate pairs: `U+D800` `U+DC3D`

----
Selection tests copied from #6274 
Where requested, these tests should be run using the  **new version 1.1** [web_context_tests keyboard](https://drive.google.com/file/d/1OhTVyN0HGgf-Jdvw6y46msfh3pwhgeCP/view?usp=sharing)

Link to the old version 1.0 [web_context_tests keyboard](https://drive.google.com/file/d/1C1I39AxR9egfuQKXRbwwyK4T9YBw6Ji-/view?usp=sharing) kept for posterity.

**GROUP_ANDROID_INAPP:** Test on Keyman for Android, in-app
**GROUP_ANDROID_SYSTEM:** Test on Keyman for Android, system keyboard (e.g. type in Browser search field)


* **TEST_ROTA** - This verifies the rota keys correctly replace characters. Keyboard  [reference](https://help.keyman.com/keyboard/xinaliq/1.1.3/xinaliq#toc-so-we-can-summarize-as-follows-) 
1. In Keyman, switch to the xinaliq keyboard.
2. From the Keyman menu, increase the text size to 57. This will make it easier to see the characters when testing in-app
3. In the text field, repeatedly press the following keys up to 5 or 6 times each, and verify the rota output

| Key                     | 1x | 2x  | 3x | 4x | 5x | 6x |
|-------------------|----|----|----|----|----|----|
| <kbd>q</kbd> | q   | q̇   | q̂   | q̂q̂ | q |      |
| <kbd>t</kbd>  |  t   |  ṫ   | t̂    | t̂t̂   | tt | t   |
| <kbd>p</kbd> | p   |  ṗ  | p̂   | p̂p̂ | p |      |
| <kbd>k</kbd> | k    | k̇   | k̂   | k̂k̂  | k |      |
| <kbd>s</kbd> | s    | ss  | ŝ    | ş    | s |      |
| <kbd>c</kbd> | c   | ċ    | ĉ    | ĉĉ   |ç |  c  |

* **TEST_BACKSPACE** - This verifies the backspace key correctly deletes 
1. With Keyman **xinaliq** keyboard selected
2. Press <kbd>k</kbd> to output `k̂k̂`
3. Press backspace
4. Verify the second combining circumflex accent is removed
5. Press backspace
6. Verify the second `k` is removed
7. Press backspace
8. Verify the first combining circumflex accent is removed
9. Press backspace
10. Verify the first `k` is removed
11. Use the globe key to switch to the **"Old Italic"** Keyman keyboard
12. Browse to the custom.html page (see setup steps)
13. On the custom.html page, copy enough text that includes the smiley emoji and the split surrogate pairs (rendered as diamonds with a question mark).
14. Paste the copied text into the text field
15. With the Old Italic keyboard, type a few SMP characters to surround the split surrogate pairs
16. Select the end of the text content
17. Hit backspace multiple times
18. Verify each backspace removes the entire character at the cursor:
    * An entire SMP character is removed and does not create a split surrogate pair (diamonds/question marks)
    * A split surrogate pair is removed
    * A normal character is removed
    * A combining diacritic is removed, leaving the base character

**TEST_BASELINE:** Using a keyboard with some complex behaviours, such as Khmer Angkor, verify that input sequences are correct. For example, type <kbd>x</kbd><kbd>E</kbd><kbd>j</kbd><kbd>m</kbd><kbd>r</kbd> (desktop: <kbd>ខ</kbd><kbd>ែ</kbd><kbd>្</kbd><kbd>ម</kbd><kbd>រ</kbd> touch: <kbd>ខ</kbd><kbd>ែ (longpress េ)</kbd><kbd>្ម (longpress ម)</kbd><kbd>រ</kbd>) and verify that the emitted output is actually `ខ្មែរ` (you may need to use a tool to view the Unicode values, which should be as in the image below).

![image](https://user-images.githubusercontent.com/4498365/153732875-71312511-7784-4ae0-8833-7f7fc23188ef.png)

**TEST_SELECTION:** Using the web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select `bc` and type <kbd>q</kbd>. The output after the final key should be `aqx`. (This is testing selection replacement with no matching keyboard rule.)

**TEST_SELECTION_2:** Using the web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select `bc` and type <kbd>p</kbd>. The output after the final key should be `aqx`. (This is testing selection replacement with a matching keyboard rule.)

**TEST_CONTEXT_BASE:** Using the web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>d</kbd>. The output after the final key should be `!`.

**TEST_CONTEXT_SELECTION:** Using the attached web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select the `x` character, and type <kbd>d</kbd>. The output after the final key should be `abcd`.

**TEST_CONTEXT_SELECTION_2:** Using the attached web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select the `x` character, and delete it with <kbd>Backspace</kbd>. Type <kbd>d</kbd>. The output after the final key should be `!`.

**TEST_CONTEXT_SELECTION_3:** Using the attached web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select the `x` character, and type <kbd>y</kbd>. Press <kbd>Backspace</kbd>, and type <kbd>d</kbd>. The output after the final key should be `!`.

**TEST_CONTEXT_SELECTION_4:** Using the attached web_context_tests keyboard, type <kbd>x</kbd><kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select the `abc` characters, and type <kbd>d</kbd>. The output after the final key should be `xdx`.

## SMP Tests

**TEST_SELECTION_SMP:** Using the web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>j</kbd>. Select `𐌁𐌂` and type <kbd>o</kbd>. The output after the final key should be `𐌀𐍈𐌈`.

**TEST_CONTEXT_BASE_SMP:** Using the web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>i</kbd>. The output after the final key should be `𐌙`.

**TEST_CONTEXT_SELECTION_SMP:** Using the attached web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>j</kbd>. Select the `𐌈` character, and type <kbd>i</kbd>. The output after the final key should be `𐌀𐌁𐌂i`.

**TEST_CONTEXT_SELECTION_SMP_2:** Using the attached web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>j</kbd>. Select the `𐌈` character, and delete it with <kbd>Backspace</kbd>. Type <kbd>i</kbd>. The output after the final key should be `𐌙`.

**TEST_CONTEXT_SELECTION_SMP_3:** Using the attached web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>j</kbd>. Select the `𐌈` character, and type <kbd>k</kbd>. Press <kbd>Backspace</kbd>, and type <kbd>i</kbd>. The output after the final key should be `𐌙`.

**TEST_CONTEXT_SELECTION_SMP_4:** Using the attached web_context_tests keyboard, type <kbd>k</kbd><kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>k</kbd>. Select the `𐌀𐌁𐌂` characters, and type <kbd>i</kbd>. The output after the final key should be `𐌰i𐌰`.
